src/mapgen_v7.cpp

```diff
@@ -483,25 +480,34 @@ void MapgenV7::carveRivers() {
 
 int MapgenV7::generateTerrain()
 {
-	int ymax = generateBaseTerrain();
+	int stone_surface_max_y = -MAP_GENERATION_LIMIT;
+	bool generate_mountain = false;
+
+	generateBaseTerrain(&stone_surface_max_y, &generate_mountain);
+
+	if ((spflags & MGV7_MOUNTAINS) && generate_mountain) {
+		int x = node_min.X;
+		int y = node_min.Y - 1;
+		int z = node_min.Z;
+		noise_mountain->perlinMap3D(x, y, z);
+		noise_mount_height->perlinMap2D(x, z);
 
-	if (spflags & MGV7_MOUNTAINS)
-		ymax = generateMountainTerrain(ymax);
+		stone_surface_max_y = generateMountainTerrain(stone_surface_max_y);
+	}
 
 	if (spflags & MGV7_RIDGES)
 		generateRidgeTerrain();
 
-	return ymax;
+	return stone_surface_max_y;
 }
 
 
-int MapgenV7::generateBaseTerrain()
+void MapgenV7::generateBaseTerrain(int *stone_surface_max_y, bool *generate_mountain)
```

Comment:
Yes will change, i thought s16 more suitable but copied the int instead thinking there was some reason for it.